src/sentry/utils/appleconnect/itunes_connect.py

```diff
@@ -3,349 +3,515 @@
+        """
+        assert self.state in [ClientState.NEW, ClientState.EXPIRED]
+        url = "https://idmsa.apple.com/appleauth/auth/signin"
+        logger.debug("POST %s", url)
+
+        start_login = self.session.post(
+            url,
+            json={
+                "accountName": username,
+                "password": password,
+                "rememberMe": True,
+            },
+            headers={
+                "X-Requested-With": "XMLHttpRequest",
+                "X-Apple-Widget-Key": self.service_key,
+                "Accept": "application/json, text/javascript",
+            },
         )
 
-
-def get_dsym_url(
-    session: Session, app_id: str, bundle_short_version: str, bundle_version: str, platform: str
-) -> Optional[str]:
-    """
-    Returns the url for a dsyms bundle. The session must be logged in.
-    :return: The url to use for downloading the dsyms bundle
-    """
-    with sentry_sdk.start_span(
-        op="itunes-dsym-url", description="Request iTunes dSYM download URL"
-    ):
-        details_url = (
-            f"https://appstoreconnect.apple.com/WebObjects/iTunesConnect.woa/ra/apps/"
-            f"{app_id}/platforms/{platform}/trains/{bundle_short_version}/builds/"
-            f"{bundle_version}/details"
+        if start_login.status_code == http.HTTPStatus.OK:
+            raise ITunesError(
+                f"Two factor auth not enabled for user, status_code={start_login.status_code}"
+            )
+        elif start_login.status_code == http.HTTPStatus.CONFLICT:
+            self.session_id = start_login.headers["x-apple-id-session-id"]
+            self._scnt = start_login.headers["scnt"]
+            self.state = ClientState.AUTH_REQUESTED
+        elif start_login.status_code == http.HTTPStatus.UNAUTHORIZED:
+            raise InvalidUsernamePasswordError
+        else:
+            raise ITunesError(f"Unexpected status code form signing: {start_login.status_code}")
+
+    def two_factor_code(self, code: str) -> None:
+        """Sends the two-factor authentication code, completing authentication.
+
+        :raises: :class:`InvalidTwoFactorAuthError`
+        """
+        assert self.state is ClientState.AUTH_REQUESTED
+        url = "https://idmsa.apple.com/appleauth/auth/verify/trusteddevice/securitycode"
+        logger.debug(f"POST {url}")
+        response = self.session.post(
+            url,
+            json={
+                "securityCode": {
+                    "code": code,
+                }
+            },
+            headers={
+                "scnt": self._scnt,
+                "X-Apple-Id-Session-Id": self.session_id,
+                "Accept": "application/json",
+                "X-Apple-Widget-Key": self.service_key,
+            },
+        )
+        if not response.ok:
+            # TODO: Make invalid code distinguishable from generic error.
+            raise InvalidAuthCodeError
+        else:
+            self.state = ClientState.AUTHENTICATED
```

Comment:
similar to `sms_code`, should this be erasing (optional) properties in the client that were initialized specifically for authentication through this method?

Reply:
hum, i thought about that but i don't think it really matters.  They don't get used anymore and that's enforced by the states.

Reply:
What would be nice is to split it up in an `ITunesAuthenticator` which at the end gives you an `ITunesClient`.  That'd probably reduce some more complexity.  But I'll leave that for a followup.

Reply:
++ on splitting this up into two different classes, one for a client that is authenticating and the other for a client that is authenticated. maybe for a follow-up PR. 

i'd love to see a quick state machine that describes this whole flow that could be added in the follow-up PR (if it happens) just to make it clear what sort of state transitions are legal and which ones aren't.